File: DSA/LC/74_search_a_2d_matrix.cpp
#include <vector>

class Solution
{
public:
  bool searchMatrix(std::vector<std::vector<int>> &matrix, int target)
  {
    int topRow = 0;
    int botRow = matrix.size() - 1;
    int midRow{};

    while (topRow <= botRow)
    {
      midRow = (botRow - topRow) / 2 + topRow;
      int firstCell = matrix[midRow][0];
      if (firstCell == target)
      {
        return true;
      }
      else if (target < firstCell)
      {
        botRow = midRow - 1;
      }
      else
      {
        topRow = midRow + 1;
      }
    }

    if (botRow == -1)
    {
      return false;
    }
    midRow = botRow;

    int leftCol = 0;
    int rightCol = matrix[midRow].size() - 1;
    int midCol{};

    while (leftCol <= rightCol)
    {
      midCol = (rightCol - leftCol) / 2 + leftCol;
      int cell = matrix[midRow][midCol];
      if (cell == target)
      {
        return true;
      }
      else if (target < cell)
      {
        rightCol = midCol - 1;
      }
      else
      {
        leftCol = midCol + 1;
      }
    }

    return false;
  }
};

File: DSA/LC/74_search_a_2d_matrix.py
from typing import List


class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        start = 0
        end = len(matrix) - 1
        while (start <= end):
            mid = (end - start) // 2 + start
            if matrix[mid][0] <= target <= matrix[mid][-1]:
                return self.binS(matrix[mid], target)
            elif target < matrix[mid][0]:
                end = mid - 1
            else:
                start = mid + 1
        return False

    def binS(self, m_list: List[int], target: int) -> bool:
        start = 0
        end = len(m_list) - 1
        while (start <= end):
            mid = (end - start) // 2 + start
            if m_list[mid] == target:
                return True
            elif target < m_list[mid]:
                end = mid - 1
            else:
                start = mid + 1
        return False
